File: src/main/java/io/github/tanguygab/tabadditions/spigot/SpigotListener.java
package io.github.tanguygab.tabadditions.spigot;

import io.github.tanguygab.tabadditions.shared.features.chat.Chat;
import me.neznamy.tab.api.placeholder.PlayerPlaceholder;
import me.neznamy.tab.shared.TAB;
import me.neznamy.tab.shared.platform.TabPlayer;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.event.player.PlayerCommandPreprocessEvent;
import org.bukkit.event.player.PlayerEvent;
import org.bukkit.event.player.PlayerToggleSneakEvent;

public class SpigotListener implements Listener {

    private final TAB tab = TAB.getInstance();

    @EventHandler
    public void onSneak(PlayerToggleSneakEvent e) {
        ((PlayerPlaceholder) tab.getPlaceholderManager().getPlaceholder("%sneak%")).updateValue(getPlayer(e),e.isSneaking());
    }


    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void onChat(AsyncPlayerChatEvent e) {
        if (!tab.getFeatureManager().isFeatureEnabled("Chat")) return;
        e.setCancelled(true);
        ((Chat)tab.getFeatureManager().getFeature("Chat")).onChat(getPlayer(e),e.getMessage());
    }

    @EventHandler
    public void onCommand(PlayerCommandPreprocessEvent e) {
        if (!tab.getFeatureManager().isFeatureEnabled("Chat")) return;
        if (((Chat)tab.getFeatureManager().getFeature("Chat")).onCommand(getPlayer(e),e.getMessage()))
            e.setCancelled(true);
    }

    private TabPlayer getPlayer(PlayerEvent e) {
        return tab.getPlayer(e.getPlayer().getUniqueId());
    }
}

File: src/main/kotlin/io/github/tanguygab/tabadditions/spigot/SpigotListener.kt
package io.github.tanguygab.tabadditions.spigot

import io.github.tanguygab.tabadditions.shared.features.chat.Chat
import me.neznamy.tab.api.placeholder.PlayerPlaceholder
import me.neznamy.tab.shared.TAB
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.event.player.PlayerEvent
import org.bukkit.event.player.PlayerToggleSneakEvent

class SpigotListener : Listener {
    private val tab = TAB.getInstance()

    @EventHandler
    fun onSneak(e: PlayerToggleSneakEvent) {
        (tab.placeholderManager.getPlaceholder("%sneak%") as PlayerPlaceholder).updateValue(
            getPlayer(e),
            e.isSneaking.toString()
        )
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onChat(e: AsyncPlayerChatEvent) {
        if (!tab.featureManager.isFeatureEnabled("Chat")) return
        e.isCancelled = true
        tab.featureManager.getFeature<Chat>("Chat").onChat(getPlayer(e), e.message)
    }

    private fun getPlayer(e: PlayerEvent) = tab.getPlayer(e.player.uniqueId)!!
}
